core/vm/operations_acl.go

```diff
@@ -242,3 +243,102 @@ func makeSelfdestructGasFn(refundsEnabled bool) gasFunc {
 	}
 	return gasFunc
 }
+
+var (
+	gasCallEIP7702         = makeCallVariantGasCallEIP7702(gasCall)
+	gasDelegateCallEIP7702 = makeCallVariantGasCallEIP7702(gasDelegateCall)
+	gasStaticCallEIP7702   = makeCallVariantGasCallEIP7702(gasStaticCall)
+	gasCallCodeEIP7702     = makeCallVariantGasCallEIP7702(gasCallCode)
+)
+
+func makeCallVariantGasCallEIP7702(oldCalculator gasFunc) gasFunc {
+	return func(evm *EVM, contract *Contract, stack *Stack, mem *Memory, memorySize uint64) (uint64, error) {
+		addr := common.Address(stack.Back(1).Bytes20())
+		// Check slot presence in the access list
+		warmAccess := evm.StateDB.AddressInAccessList(addr)
+		// The WarmStorageReadCostEIP2929 (100) is already deducted in the form of a constant cost, so
+		// the cost to charge for cold access, if any, is Cold - Warm
+		coldCost := params.ColdAccountAccessCostEIP2929 - params.WarmStorageReadCostEIP2929
+		if !warmAccess {
+			evm.StateDB.AddAddressToAccessList(addr)
+			// Charge the remaining difference here already, to correctly calculate available
+			// gas for call
+			if !contract.UseGas(coldCost, evm.Config.Tracer, tracing.GasChangeCallStorageColdAccess) {
+				return 0, ErrOutOfGas
+			}
+		}
+
+		// Check if code is a delegation and if so, charge for resolution.
+		if addr, ok := types.ParseDelegation(evm.StateDB.GetCode(addr)); ok {
+			var cost uint64
+			if evm.StateDB.AddressInAccessList(addr) {
+				cost += params.WarmStorageReadCostEIP2929
+			} else {
+				evm.StateDB.AddAddressToAccessList(addr)
+				cost += params.ColdAccountAccessCostEIP2929
+			}
+			if !contract.UseGas(cost, evm.Config.Tracer, tracing.GasChangeCallStorageColdAccess) {
+				return 0, ErrOutOfGas
+			}
+			coldCost += cost
```

Comment:
If the `addr` is already warmed; the additional delegation gas charging will be skipped

Reply:
Looking into this more, this is expected behavior? There are a few warming patterns to consider:

* cold-cold: the addr and target are cold, therefore both need to be added to access list and charged
* cold-warm: the addr is cold, but its target is warm. This can happen for many reasons, but imagine many people delegate to the same target. In this case we need to charge for warming addr, and then need to tack on a warm access cost for the delegation target. That happens on L275.
* warm-warm: both addr and its target are warm. This payment again happens on L275.
* warm-cold: the addr is warm but it's target is not. I think right now only possible via SELFDESTRUCT, but after devnet-5 will be much more common. We charge for cold costs of the delegation target on L280 again.

So AFAICT this is the correct behavior currently. Let me know if you see something wrong in this reasoning.

Reply:
For warm-cold case, yes the cold cost is charged on L280, but it's not correct.

Look at the original code of `makeCallVariantGasCallEIP2929`

```
func makeCallVariantGasCallEIP2929(oldCalculator gasFunc, addressPosition int) gasFunc {
	return func(evm *EVM, contract *Contract, stack *Stack, mem *Memory, memorySize uint64) (uint64, error) {
		addr := common.Address(stack.Back(addressPosition).Bytes20())
		// Check slot presence in the access list
		warmAccess := evm.StateDB.AddressInAccessList(addr)
		// The WarmStorageReadCostEIP2929 (100) is already deducted in the form of a constant cost, so
		// the cost to charge for cold access, if any, is Cold - Warm
		coldCost := params.ColdAccountAccessCostEIP2929 - params.WarmStorageReadCostEIP2929
		if !warmAccess {
			evm.StateDB.AddAddressToAccessList(addr)
			// Charge the remaining difference here already, to correctly calculate available
			// gas for call
			if !contract.UseGas(coldCost, evm.Config.Tracer, tracing.GasChangeCallStorageColdAccess) {
				return 0, ErrOutOfGas
			}
		}
		// Now call the old calculator, which takes into account
		// - create new account
		// - transfer value
		// - memory expansion
		// - 63/64ths rule
		gas, err := oldCalculator(evm, contract, stack, mem, memorySize)
		if warmAccess || err != nil {
			return gas, err
		}
		// In case of a cold access, we temporarily add the cold charge back, and also
		// add it to the returned gas. By adding it to the return, it will be charged
		// outside of this function, as part of the dynamic gas, and that will make it
		// also become correctly reported to tracers.
		contract.Gas += coldCost

		var overflow bool
		if gas, overflow = math.SafeAdd(gas, coldCost); overflow {
			return 0, ErrGasUintOverflow
		}
		return gas, nil
	}
}
```

You will find that it's a bit tricky here. What it does it:

- if the address is warmed, fallback to `oldCalculator`;
- if the address is not warmed, charge the cold cost, invoke the `oldCalculate` with the **updated remaining gas by deducting the cold cost**, then **add the charged cold cost back on L184**, add the cold cost to the returned gas
- the returned gas amount will be regarded as the dynamic gas cost and be deducted outside

The reason behind is in `oldCalculate`, we have this 63/64ths rule requiring the accurate remaining gas for calculation; therefore code cost pre-deduction is needed. cc @holiman please correct me if I am wrong here.

So, the same logic for 7702, no matter the address is not warmed, or the delegation is not warmed, charge it temporarily, invoke the oldCalculator, then add the cost cost back. Apparently it's broken with the current implementation in case `warm-cold`

Reply:
> So, the same logic for 7702, no matter the address is not warmed, or the delegation is not warmed, charge it temporarily, invoke the oldCalculator, then add the cost cost back. Apparently it's broken with the current implementation in case `warm-cold`

I'm comparing the two implementations side by side, `makeCallVariantGasCallEIP2929` vs `makeCallVariantGasCallEIP7702`, and it looks to me like he is doing exactly what you propose he should be doing. Not sure what I'm not seeing here. Related: would be neat if we can come up with a statetest for it

Reply:
ah, now I "see" it
```golang
		if warmAccess || err != nil {
			return gas, err
		}
```

So, the problem is the "early-ish return", if `warmAccess` is `true`.

Reply:
As I see it, something like this should be more correct: 
```diff
diff --git a/core/vm/operations_acl.go b/core/vm/operations_acl.go
index 4245e7ae4a..d7a599938c 100644
--- a/core/vm/operations_acl.go
+++ b/core/vm/operations_acl.go
@@ -254,12 +254,13 @@ var (
 func makeCallVariantGasCallEIP7702(oldCalculator gasFunc) gasFunc {
 	return func(evm *EVM, contract *Contract, stack *Stack, mem *Memory, memorySize uint64) (uint64, error) {
 		addr := common.Address(stack.Back(1).Bytes20())
+		var coldCost uint64
 		// Check slot presence in the access list
 		warmAccess := evm.StateDB.AddressInAccessList(addr)
-		// The WarmStorageReadCostEIP2929 (100) is already deducted in the form of a constant cost, so
-		// the cost to charge for cold access, if any, is Cold - Warm
-		coldCost := params.ColdAccountAccessCostEIP2929 - params.WarmStorageReadCostEIP2929
 		if !warmAccess {
+			// The WarmStorageReadCostEIP2929 (100) is already deducted in the form of a constant cost, so
+			// the cost to charge for cold access, if any, is Cold - Warm
+			coldCost += (params.ColdAccountAccessCostEIP2929 - params.WarmStorageReadCostEIP2929)
 			evm.StateDB.AddAddressToAccessList(addr)
 			// Charge the remaining difference here already, to correctly calculate available
 			// gas for call
@@ -288,7 +289,7 @@ func makeCallVariantGasCallEIP7702(oldCalculator gasFunc) gasFunc {
 		// - memory expansion
 		// - 63/64ths rule
 		gas, err := oldCalculator(evm, contract, stack, mem, memorySize)
-		if warmAccess || err != nil {
+		if coldCost == 0 || err != nil {
 			return gas, err
 		}
 		// In case of a cold access, we temporarily add the cold charge back, and also
```

But, again, I'm more interested in finding a statetest to trigger the difference, than nailing down the best code right now

Reply:
Yep, something like this is much better.

Reply:
So the way I understand it: the dynamic call gas variant functions use `contract.UseGas` to determine if the contract can pay then later add back the cost so that it can return the total sum of dynamically used gas to be deducted in the interpreter.

Given this, I think the behavior will be correct w.r.t. gas used, but it may cause an issue with tracers. The reason being, the gas is charged as expected, and it is just not refunded in the case of a cold access. So since AFAIK it's correctly consumed, I don't think it will be possible to create a state test that creates an issue here. @holiman please LMK if you found otherwise.

That said, yeah it is a good point and Martin's code change recommendation appears more correct so I think we can make this change.

Given how much testing this code has seen and it was a bit tricky to get right - I want to first make sure the tests for devnet-5 are passing then make the change so that we're not also hunting for new issues introduced here.

Reply:
Yeah, i would also prefer to go with the change, at least easier to understand, more correct for tracer, and less likely to be buggy.

Reply:
Since there are no filled tests for devnet-5 yet, haven't been able to verify correctness, but I did implement the change in [bc0e65a](https://github.com/ethereum/go-ethereum/pull/30078/commits/bc0e65a7f94e2e7585ca730a4366b8355cb59b68).

I was able to write a test for the tracer to make sure cost is reported correctly, so I think the change is okay.